driver/main.c

```diff
@@ -343,13 +535,60 @@ static void check_remove_consumer(struct ppm_consumer_t *consumer, int remove_fr
 	}
 }
 
+static void set_consumer_tracepoints(struct ppm_consumer_t *consumer, u32 tp_set)
+{
+	int i;
+	u32 stored_tp_refs[TP_VAL_MAX];
+
+	memcpy(stored_tp_refs, g_tracepoints_refs, sizeof(g_tracepoints_refs));
+	if (force_tp_set(consumer, tp_set, TP_VAL_MAX) == 0)
+	{
+		// no error, we can store tp_set
+		consumer->tracepoints_attached = tp_set;
+	}
+	else
+	{
+		/*
+		 * In case of error, we check the stored refs
+		 * to see if anything changed;
+		 * then we store properly attached
+		 * tracepoints bits for the consumer.
+		 */
+		for(i = 0; i < TP_VAL_MAX; i++)
+		{
+			if(stored_tp_refs[i] != g_tracepoints_refs[i])
```

Comment:
Be as careful as possible to avoid any desync: in case of error, check which stored ref was changed and update consumer tp mask as needed.